integration/yarn/src/main/java/tachyon/yarn/ApplicationMaster.java

```diff
@@ -178,7 +202,24 @@ public void start() throws IOException, YarnException {
   public void requestContainers() throws Exception {
     requestMasterContainer();
 
-    requestWorkerContainers();
+    // Wait until all Tachyon worker containers have been allocated
+    int round = 0;
+    synchronized (mWorkerHosts) {
```

Comment:
this code block is the core where control flow happens. Currently there are at least three variables affecting the control flow: `mWorkerHosts.size()`, `round` and `mOutstandingWorkerContainerRequests`, and there are different (callback) threads also involved. It costs me quite some time to fully understand and verify the logic here. I am wondering if we could either (1) simply the logic , or (2) have more javadoc describing the control flow? Otherwise the interaction seems fairly complicated and hard to reason if we could get into deadlock.